langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/FunctionCallHelper.java

```diff
@@ -0,0 +1,130 @@
+            case "boolean":
+                return Type.BOOLEAN;
+            case "number":
+                return Type.NUMBER;
+            case "array":
+                return Type.ARRAY;
+            case "object":
+                return Type.OBJECT;
+            default:
+                return Type.TYPE_UNSPECIFIED;
+        }
+    }
+
+    static FunctionCall fromToolExecutionRequest(ToolExecutionRequest toolExecutionRequest) {
+        FunctionCall.Builder fnCallBuilder = FunctionCall.newBuilder()
+            .setName(toolExecutionRequest.name());
+
+        Struct.Builder structBuilder = Struct.newBuilder();
+        try {
+            JsonFormat.parser().merge(toolExecutionRequest.arguments(), structBuilder);
+        } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException(e);
+        }
+        Struct argsStruct = structBuilder.build();
+        fnCallBuilder.setArgs(argsStruct);
+
+        return fnCallBuilder.build();
+    }
+
+    static ToolExecutionRequest fromFunctionCall(FunctionCall functionCall) {
+        ToolExecutionRequest.Builder builder = ToolExecutionRequest.builder()
+            .name(functionCall.getName());
+
+        //TODO: can we make the mapping more transparently/automatically?
+
+        Map<String, Object> callArgsMap = new HashMap<>();
+        Struct callArgs = functionCall.getArgs();
+        Map<String, Value> callArgsFieldsMap = callArgs.getFieldsMap();
+        callArgsFieldsMap.forEach((key, value) -> callArgsMap.put(key, unwrapProtoValue(value)));
+
+        String serializedArgsMap = ToolExecutionRequestUtil.GSON.toJson(callArgsMap);
+        builder.arguments(serializedArgsMap);
+
+        return builder.build();
+    }
+
+    static Object unwrapProtoValue(Value value) {
+        Object unwrappedValue;
+        switch (value.getKindCase()) {
+            case NUMBER_VALUE:
+                unwrappedValue = value.getNumberValue();
+                break;
+            case STRING_VALUE:
+                unwrappedValue = value.getStringValue();
+                break;
+            case BOOL_VALUE:
+                unwrappedValue = value.getBoolValue();
+                break;
+            case STRUCT_VALUE:
+                HashMap<String, Object> mapForStruct = new HashMap<>();
+                value.getStructValue().getFieldsMap().forEach((key, val) -> mapForStruct.put(key, unwrapProtoValue(val)));
+                unwrappedValue = mapForStruct;
+                break;
+            case LIST_VALUE:
+                unwrappedValue = value.getListValue().getValuesList().stream().map(FunctionCallHelper::unwrapProtoValue).collect(Collectors.toList());
+                break;
+            default: // NULL_VALUE, KIND_NOT_SET, and default
+                unwrappedValue = null;
+                break;
+        }
+        return unwrappedValue;
+    }
+
+    static Tool convertToolSpecifications(List<ToolSpecification> toolSpecifications) {
```

Comment:
Shouldn't it be a _list_ of `Tool`s from a list of `ToolSpecification`s?

Reply:
That's a good question!
If I understand correctly, a `Tool` is a set of functions, and we can pass one or more tools.
But if there are several functions, should I create a tool per function?
What's the logic with the organisations of functions as set of tools?

Reply:
In my understanding 1 function = 1 tool. Tool can also be something else than function (at least in OpenAI nomenclature).

Reply:
I had a look at the data models of OpenAI and Gemini.

* In OpenAI's [doc](https://platform.openai.com/docs/assistants/tools/defining-functions), there's indeed one top-level `tools` array, which can contain _tools_ which can be function declarations.
* But in Gemini's [doc](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling#function-calling-samples), the `tools` array can contain tools which can be a `function_declarations` that can in turn contain an array of function declarations.

Currently, indeed, I create one tool that contains all the function declarations.
It seems that's what most code samples I've seen are doing.
But I don't know if it makes a difference or not if I created one tool per function.
I think I'll stick to this approach for now, for lack of a better understanding of the internals.

Reply:
Got confirmation that there is currently only one type of tool which is a set of functions.
There may be other tool types later on, but for now, it makes sense to just return one `Tool` that contains an array of `function_declarations`.